Use pyproject.toml instead of setup.cfg, pytest.ini & requirements.txt and other Python related config files

stv0g opened the pull request:
Modern Python is slowly adopting `pyproject.toml` as a single place to define a Python package and related tool settings like black for formatting, flake8 for linting and pytest for testing.

Lets try to move as much of the Python-related project configuration to this file and unclutter the root of the repo.

This PR does:
- Merge `requirements*.txt`, `pytest.ini` and `setup.cfg` files into `pyproject.toml`
- Update shell scripts and Dockerfiles to use `pyproject.toml`
- Introduce a pre-commit configuration file
- Add a GitHub actions workflow for running pre-commit also in the CI
- Apply/fix pre-commit checks to all files in the repo:
  - Harmonize file endings (single empty line)
  - Remove trailing whitespaces
  - Apply formatting of all Python code (also in Jupyter notebooks)
- Remove obsolete CI jobs from `.gitlab-ci.yml`

steffen-vogel-opalrt committed:
python: Harmonize comment style

Signed-off-by: Steffen Vogel <steffen.vogel@opal-rt.com>

steffen-vogel-opalrt committed:
Move pytest configuration to pyproject.toml
python: pick 266daa8f wip: villas: Minor fixes and tweaks for VillasInterface

Signed-off-by: Steffen Vogel <steffen.vogel@opal-rt.com>

steffen-vogel-opalrt committed:
Add pre-commit

Signed-off-by: Steffen Vogel <steffen.vogel@opal-rt.com>

steffen-vogel-opalrt committed:
python: Use pyproject.toml instead of setup.cfg and requirements.txt

Signed-off-by: Steffen Vogel <steffen.vogel@opal-rt.com>

stv0g commented:
I am interested in your opinions @leonardocarreras and @m-mirz.
What do you think about consolidate all the Python related project files in `pyproject.toml`

m-mirz commented:
Good idea 👍

leonardocarreras commented:
I think it is a great idea 🙌🏻

stv0g committed:
Fixup whitespaces

Remove trailing whitespaces
Harmonize file endings to single empty line

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Fix editorconfig for new clang-format style

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Format all Python code and Jupyter notebooks with black formatter

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Fix formatting of JOSN files

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Update pre-commit config

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Reformat code with clang-format

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
ci: Remove obsolete jobs from gitlab-ci.yml

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Fix some capitalization

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g commented:
This is ready for review :)

I've also added pre-commit as a tool to perform several checks against changes which are staged for commit. Such as removing trailing whitespaces, checking JSON, YAML and XML syntax and Python / C++ code formatting.

While doing so, I realized that much of the Python code was ugly formatted.
We are now using the de-facto standard Python formatter [Black](https://github.com/psf/black).
It even can format Python code with in Jupyter notebooks :)

stv0g commented:
@m-mirz I guess the SonarCloud warnings are false-positives. Its at least not code which I've touched.

stv0g commented:
Hi @martinmoraga @leonardocarreras @dinkelbachjan,

Markus has reviewed and approved this change. However, we would like to find more approvals before going ahead and merging it.

Do you have a second to look at this change?

martinmoraga commented on pyproject.toml:
```diff
@@ -1,9 +1,67 @@
+description = "Dynamic real-time power system simulator"
+readme =  {file = "README.md", content-type = "text/markdown"}
+authors = [
+    {name = "The DPsim Authors", email = "mmirz@eonerc.rwth-aachen.de"},
```
@m-mirz Is this email address still valid?

m-mirz replied:
@martinmoraga no it is not. @stv0g It should be markus.mirz@rwth-aachen.de

stv0g replied:
I will fix it.

leonardocarreras commented on setup.py:
```diff
@@ -1,82 +1,64 @@
+        cfg = "Debug" if self.debug else "Release"
+        print("building CMake extension in %s configuration" % cfg)
 
         cmake_args = [
```
Maybe I am doing something wrong, but with `pip install .` it does not work because `pybind` is disabled, and I need to `pip install . --config-settings="-DFETCH_PYBIND=ON"`... Should we have it active by default?

leonardocarreras commented on packaging/Shell/install-fedora-deps-dev.sh:
```diff
@@ -47,43 +47,43 @@ dnf -y debuginfo-install \
-pip3 install -r requirements.txt
+pip3 install --upgrade setuptools
+pip3 install --upgrade wheel
+pip3 install .[test,docs]
```
Here looks like we are doing a different thing, not only installing dependencies but also the package in the currently active folder (dpsim). Is this the intended behaviour?

m-mirz replied:
In the dev container, the dpsim package should not be installed... only the dependencies.

leonardocarreras commented on packaging/Shell/install-fedora-deps.sh:
```diff
@@ -22,18 +22,18 @@ dnf -y install \
-pip3 install -r requirements.txt
+pip3 install --upgrade setuptools
+pip3 install --upgrade wheel
+pip3 install .
```
Any particular reason for the difference here about only `pip3 install .` vs `.[test,docs]`  in the `packaging/Shell/install-fedora-deps-dev.sh`

m-mirz replied:
@stv0g Doesn't this install the dpsim package as well and not just deps?

leonardocarreras commented on packaging/Shell/install-ubuntu-deps.sh:
```diff
@@ -24,8 +24,7 @@ apt-get -y install \
 
-pip3 install --user -r requirements.txt
-pip3 install --user -r requirements-jupyter.txt
+pip3 install --user .
```
Any particular reason for the difference here about only pip3 install . vs .[test,docs] in the packaging/Shell/install-fedora-deps-dev.sh

m-mirz replied:
@stv0g Here I would also expect only the installation of deps and not the dpsim package.

leonardocarreras commented on packaging/Shell/install-fedora-deps.sh:
```diff
@@ -56,22 +56,22 @@ sudo ldconfig
-	rm -rf /tmp/libcimpp
+    rm -rf /tmp/libcimpp
 
 # Install VILLASnode from source
```
I think it may be the case that L36 already installed `villas-node `as dependency due to it being listed in the `.toml`, this may be also the case for the similar shell files and even the dockerfiles

stv0g replied:
L36 only installed the [VILLASnode Python package](https://pypi.org/project/villas-node/) but noch actually compile/build VILLASnode's C++ code

leonardocarreras replied:
Thanks! Then disregard this too

leonardocarreras commented on setup.py:
```diff
@@ -1,82 +1,64 @@
+        cfg = "Debug" if self.debug else "Release"
+        print("building CMake extension in %s configuration" % cfg)
 
         cmake_args = [
```
When only using the normal compilation, this happens

`subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'dpsimpy', '--', '-j4']' returned non-zero exit status 2.`

Should we enable `-DFETCH_PYBIND` here or is something only related to my system?

stv0g replied:
Good question.. I would say its depending on the environment where you build DPsim and which dependencies are available via the systems package manager.

I think we should not change the flags in this PR. If its an issue, its unrelated to the original goal of this PR and hence should be handled in a separate one.

leonardocarreras commented on packaging/Docker/Dockerfile.dev-minimal:
```diff
@@ -22,8 +22,8 @@ RUN dnf --refresh -y install \
-RUN pip3 install -r requirements.txt
+RUN pip3 install .[test,docs]
 
 EXPOSE 8888
```
Why do we expose 8888 here? LIkely left from some jupyter

m-mirz replied:
This is intended to be able to start and use jupyterlab from within the dev container.

stv0g replied:
I think its not required. `EXPOSE` directives are not mandatory for accessing the container. Its only used by an "expose all ports" setting. 

I will remove it here as we dont even install Jupyter in this Dockerfile.
It might make sense in the Dockerfile in which. Jupyter is actually installed

leonardocarreras reviewed (dismissed):
Hi @stv0g! Thanks for the extra formatting, I think the code will look a lot more standardized now... From the side of the code it looks good to me in general, I would like some more opinions still on the _how we do_...

leonardocarreras commented:
Side topic: Apparently, the more changes are here, the more **Sonar** complains and demands higher quality from reformatted existing code...

stv0g committed:
Fix Markus' mail address

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Remove broken Jupyter installation from Dockerfile.rpm

It refers to a file named requirements-jupyter.txt which
does not exist for quite some time now..

Also the file was broken by a missing line continuation..

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Remove superfuous EXPOSE directives from Dockerfiles

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

stv0g committed:
Cleanup Dockerfiles

Signed-off-by: Steffen Vogel <post@steffenvogel.de>

m-mirz commented:
@stv0g @leonardocarreras I think this can be merged after the question regarding pip install have been resolved.

stv0g commented:
I converted this PR to a draft, since I am splitting it up into smaller changesets and new PRs.

Once, everthing is merged, I will close this one.

leonardocarreras commented:
Thanks so much for all the ideas and implementation @stv0g. I am closing this PR after last PR #485 was merged. I understand that the rest of the changes were integrated by you in previous steps. If you think that something is still missing, please feel free to reopen.
